Add function to delete all objects across all entities in data model

BendyWalker opened the pull request:
Implements feature referenced in #9.

Commit:
Add batchDeleteAllEntities() function

blork commented:
Seems like a good addition. Only thing stopping me is the lack of tests. It definitely works, right? It's been used somewhere?

blork reviewed (changes requested):
Seems like a good addition. Only thing stopping me is the lack of tests. It definitely works, right? It's been used somewhere?

BendyWalker commented:
> Seems like a good addition. Only thing stopping me is the lack of tests. It definitely works, right? It's been used somewhere?

@blork Yeah, it's being used on logout to wipe user data in one of our internal projects. I can do some further investigation into if any testing would be possible.

djyates commented on PeakCoreData/Helpers/NSManagedObjectContext.swift:
```diff
@@ -44,4 +44,29 @@ extension NSManagedObjectContext {
+     
+     - parameter context:   The context to use.
+     */
+    static func batchDeleteAllEntities() {
```
You've made this a static function on `NSManagedObjectContext`, which means it can't access instance member `persistentStoreCoordinator`. Have you run this on some data to test it works?

BendyWalker replied:
This was a copy-paste error. This function should not be static.

djyates reviewed (changes requested):
Not a fan of the duplication here as this method is almost exactly the same as the `batchDelete(in:)` method on `ManagedObjectType`. Wondering if we could move this to an extension on `NSEntityDescription` so it can be called from both places. 

Also, given this is a long running process it might make sense to create a `CoreDataOperation` that does this to enforce doing this on a background thread. Would also make it harder to run by accident!

As unit tests aren't possible, could you add something to the example project that runs this to show it in action.

blork commented:
Not sure if we should create a specific operation for it. You can always wrap it in a `CoreDataOperation` yourself, if you need it. But then we do have things like `CoreDataBatchImportOperation` for potentially slow tasks. So I'm torn. Perhaps add the operation, as long as it can still be called manually if you're using some other backgrounding method.

Commit:
Update pods

Commit:
Extract batch delete logic to function on NSEntityDescription

Commit:
Add Person entity and create attendee relationship to Event

Commit:
Display list of attendees upon tapping on an event

Commit:
Publicise batchDeleteAllEntities function

Commit:
Call batchDeleteAllEntities() from EventTableViewController

Commit:
Wrap batchDeleteAllEntities in operation

Commit:
Add missing files to Xcode pbxproj

Commit:
Tweak documentation on a couple of functions

Commit:
Remove dedicated batch delete operation

Commit:
Add optional 'mergingInto:mergeContexts' parameter to batch delete functions

BendyWalker commented:
I've extracted the common batch delete logic to a function on `NSEntityDescription` and updated the example project to demonstrate it working. I briefly implemented an operation, but decided it was a little too specific to belong in the framework. The current implementation is flexible enough to be called on the view context, background context or wrapped up and called in a `CoreDataOperation`.

djyates commented on Example/PeakCoreDataExample/PeakCoreDataExample/EventsTableViewController.swift:
```diff
@@ -53,20 +53,19 @@ class EventsTableViewController: UITableViewController, PersistentContainerSetta
     
     @IBAction func deleteButtonTapped(_ sender: Any) {
-        Event.batchDelete(in: viewContext)
+        persistentContainer.performBackgroundTask { backgroundContext in
```
Could you update the equivalent on `EventCollectionViewController`, which I'm guessing doesn't work now.

BendyWalker replied:
Oops, forgot all about that 🤦‍♂ Arguably, that logic should be shared between the two, right? Alas, it's only an example project.

BendyWalker replied:
Copied the deletion logic to the collection view controller in https://github.com/3squared/PeakCoreData/pull/11/commits/86d5cd5d81219a68d1af0f2e0cb1df9a8a66ab61.

djyates commented on Example/PeakCoreDataExample/PeakCoreDataExample/EventDetailViewController.swift:
```diff
@@ -46,4 +50,20 @@ class EventDetailViewController: UIViewController, PersistentContainerSettable {
+    
+    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: AttendeeTableViewCell.cellIdentifier, for: indexPath) as! AttendeeTableViewCell
+        let attendee = event.attendees?.allObjects[indexPath.row] as! Person
```
I know this isn't particularly relevant, but as `attendees` is a Set this will display the attendees in a random order each time. Better to fetch the attendees that are related to the passed in event with a sort descriptor.

BendyWalker replied:
Changed the display of attendees to use data source in 97fec7b.

Commit:
Update authors in podspec

Commit:
Update deletion logic in EventsCollectionViewController

Commit:
Change EventDetailViewController to data source implementation

BendyWalker commented:
What's the next step for this pull request?
